catalog/lease: populate SessionID for historical leases

fqazi opened the pull request:
Previously, we were seeing intermittent panics in scenarios where
historical descriptors were accessed followed by reads of newer
descriptor versions. This would happen because the SessionID on
historical versions would be blank, causing our checks for detecting
changes to fails. To address this, this patch always populates the
session ID inside the descriptor versions.

Fixes: #141877
Fixes: #141876
Fixes: #141855

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141928)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141928)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141928#-OJsVFfg8DTrzsyWSM8m:-OJsVFfg8DTrzsyWSM8n:b6iwckl) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a5c22854295734f8ea6c1328b8bef03a233001/pkg/sql/catalog/lease/descriptor_version_state.go#L115)):*
> ```Go
> 	s.mu.Lock()
> 	defer s.mu.Unlock()
> 	// Historical leases will never have sessions.
> ```

why do historical leases not have sessions? (just curious for my own understanding; can you link to the code where this happens?) can we change it so they do?

the comment on `session` should be updated to describe why this is the case. also, that comment still references "dual write" mode, so it's outdated.

https://github.com/cockroachdb/cockroach/blob/aeb5a7b7604bcce3c17bb3d107f0b2aa0724227b/pkg/sql/catalog/lease/descriptor_version_state.go#L63-L66


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141928)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141928#-OJsVFfg8DTrzsyWSM8m:-OJsX-PN11amkMkVr2sA:b-j1z2bu) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a5c22854295734f8ea6c1328b8bef03a233001/pkg/sql/catalog/lease/descriptor_version_state.go#L115)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why do historical leases not have sessions? (just curious for my own understanding; can you link to the code where this happens?) can we change it so they do?

the comment on `session` should be updated to describe why this is the case. also, that comment still references "dual write" mode, so it's outdated.

https://github.com/cockroachdb/cockroach/blob/aeb5a7b7604bcce3c17bb3d107f0b2aa0724227b/pkg/sql/catalog/lease/descriptor_version_state.go#L63-L66
</blockquote></details>

When I added session based leasing this never got populated for historical descriptors, because no stored lease existed. https://github.com/cockroachdb/cockroach/blob/b9c6bb6a0570d The code path is: 1022b34c0be699cb1c9cf10126e/pkg/sql/catalog/lease/lease.go#L1470

If you start having code paths reading older versions via AOST or just the timestamp that you currently have this code will end up inserting leases without a SessionID. 

I'll modify it to always populated a session ID instead.


<!-- Sent from Reviewable.io -->

fqazi committed:
catalog/lease: populate SessionID for historical leases

Previously, we were seeing intermittent panics in scenarios where
historical descriptors were accessed followed by reads of newer
descriptor versions. This would happen because the SessionID on
historical versions would be blank, causing our checks for detecting
changes to fails. To address this, this patch always populates the
session ID inside the descriptor versions.

Fixes: #141877
Fixes: #141876
Fixes: #141855

Release note: None

fqazi commented:
@rafiss RFAL

rafiss reviewed (approved):
lgtm! this looks like a cleaner fix than just a defensive nil-check

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141928#-:-OJt2rZt5zfkhPUK_een:bbjywh0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss TFTR!

bors r+
